### PR TITLE
Fix the routing for non-tree-spawn launch

### PR DESCRIPTION
**Pull request by rhc54**
The prted should send directly back to the HNP in this scenario. Nidmap needs to update the routing tree when we update the map as the number of daemons may have changed.

### CODE REVIEW

**Comment by rhc54** [comment]
@naughtont3 This should resolve the problem you were chasing.

**Comment by rhc54** [comment]
bot:ibm:retest

File: src/util/nidmap.c [hunk 1]
```diff
@@ -422,6 +422,9 @@ int prte_util_decode_nidmap(pmix_data_buffer_t *buf)
         prte_rml_compute_routing_tree();
     }
 
+    /* update the routing tree */
+    prte_rml_compute_routing_tree();
```
Comment (naughtont3) [inline src/util/nidmap.c:426]: @rhc54 Should the additional `prte_rml_compute_routing_tree()` on line 422 be dropped, so not compute routing tree twice?

**Comment by naughtont3** [comment]
@rhc54 still seeing odd behaviour, also seems like in the plm=ssh route with rml_base_radix=1 if I enable daemon debugs, it looks like we see too many Daemons being launched.

Instrumented debug prints to show PIDs and see that it grows with number of nodes.
```
node0:$ prte   --prtemca plm ssh --prtemca ras ^slurm  --prtemca rml_base_radix 1 --debug-daemons    --hostfile hosts4 
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33300) Daemon was launched on node1 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33366) Daemon was launched on node2 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33372) Daemon was launched on node2 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33466) Daemon was launched on node3 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33468) Daemon was launched on node3 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33477) Daemon was launched on node3 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (33475) Daemon was launched on node3 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37701) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37803) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37807) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37810) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37812) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37814) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37818) Daemon was launched on node4 - beginning to initialize
[../../../../../../3rd-party/prrte/src/tools/prted/prted.c:324] (37819) Daemon was launched on node4 - beginning to initialize
DVM ready
```

File: src/util/nidmap.c (hunk 1 above)
Comment (rhc54) [reply src/util/nidmap.c:426]: yeah - no point in doing it twice!

**Comment by rhc54** [comment]
Looks like a different problem is impacting tree-spawn. I'll have to look at that one. I was only addressing the no-tree-spawn case.

**Comment by rhc54** [comment]
Hmmm...looks like I still don't have the no-tree-spawn case quite right either. Will need to ponder a bit more.

**Comment by naughtont3** [comment]
@rhc54 did  a quick check with latest changes here and still seeing that the OOB_SEND is failing when it tries to send.  Attaching some logfiles that might be useful to you. I'm still trying to fully grok the traces.  But looks to be same with plm=slurm and plm=ssh+notree-spawn.  See msgs like `UNABLE TO SEND MESSAGE TO`.  And in this case, it is tasks sending to the HNP, but any task lower in the tree, i.e., not directly connected to [0,0], will fail.

My test is 4 nodes and rml_base_radix=1.
```
[node3:52122] [prte-node0-1251439@0,3] UNABLE TO SEND MESSAGE TO [prte-node0-1251439@0,0] TAG 10: No OOB path to target
```
- [slurm-notree-norouteSendError.txt](https://github.com/openpmix/prrte/files/10902163/slurm-notree-norouteSendError.txt)
- [ssh-notree-norouteSendError.txt](https://github.com/openpmix/prrte/files/10902165/ssh-notree-norouteSendError.txt)

**Comment by rhc54** [comment]
Yeah, I know - still working on it. It's a nasty tangle right now. I can explain better over the phone sometime if you like.

**Comment by rhc54** [comment]
Okay, this all looks good now. I've left it as a bunch of isolated commits so you can see what I had to do - I cleaned up a few things along the way as I encountered them (e.g., prte_info). One you have had a chance to take a look, I'll squash this all down to a more reasonable set of commits.

**Comment by rhc54** [comment]
Quick summary of the problem: there was never anything wrong with the routing tree. The problem was a typo in the code where we stored the HNP's URI - it was mistakenly overwriting the prted's own URI. This meant that the wireup info sent to the daemons after launch included the HNP's URI for every prted, meaning that no prted could communicate to its parent as it had the wrong URI.

Once that was fixed, the problem became that the prted wasn't storing its URI in the PMIx server, and thus we weren't able to retrieve it when we wanted to send that info to the HNP. This was corrected by simply adding the code back (unsure why it was removed in the first place).

Final part of the problem was just to tell the OOB that if it didn't know the URI on the next hop up the tree towards the HNP, then it should just send the message directly to the HNP. This ensure that the daemons could "phone home" during initial startup.

The rest of the changes address other issues I encountered along the way. mpirun was simply hanging if one or more daemons failed to report back (i.e., failed to start) - needed to tell it to exit if communications fail (since we have no way of ordering remaining daemons to terminate - at least, not yet).

prte_info wasn't showing the rml, routed and other entries if no cmd line options were provided as those entries don't have MCA components - so I have it now just reporting the name and "no components" so at least you know it exists.

**Comment by rhc54** [comment]
bot:ibm:retest

**Comment by naughtont3** [comment]
@rhc54 thanks!  i did a quick spot check on my local devel box and this seems to have fixed the tiny reproducer case I was doing with `rml_base_radix=1`.  I need to do a cleanup and retest to ensure not still getting tear down and works on full size slurm system (>65 nodes).

```
node0:$ salloc -N 4
node0:$ prte --prtemca rml_base_radix 1  --debug-daemons 
Daemon was launched on node1 - beginning to initialize
Daemon was launched on node3 - beginning to initialize
Daemon was launched on node4 - beginning to initialize
Daemon was launched on node2 - beginning to initialize
DVM ready
```

**Pull request merged**